src/image.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Github-runner-image-builder-operator image relation observer."""
+
+import logging
+from typing import TypedDict
+
+import ops
+
+import builder
+import charm_utils
+import state
+
+logger = logging.getLogger(__name__)
+
+IMAGE_RELATION = "image"
+
+
+class ImageRelationData(TypedDict):
+    """Relation data for providing image ID.
+
+    Attributes:
+        id: The latest image ID to provide.
+    """
+
+    id: str
+
+
+class Observer(ops.Object):
+    """The image relation observer."""
+
+    def __init__(self, charm: ops.CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "image-observer")
+        self.charm = charm
+
+        charm.framework.observe(
+            charm.on[IMAGE_RELATION].relation_joined, self._on_image_relation_joined
+        )
+
+    @charm_utils.block_if_invalid_config(defer=False)
+    def _on_image_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
+        """Handle the image relation joined event."""
+        build_config = state.BuilderRunConfig.from_charm(charm=self.charm)
+        image_id = builder.get_latest_image(
```

Comment:
Is there any action that the user could take in case of GetLatestImageError?

Reply:
I believe they would need to then investigate the error on why the application has raised an error - e.g. network error perhaps? (The application will return an empty string if no image was available / found :D )
Currently, there is no action that the user can take on the charm side to resolve this unfortunately, hence the fall-through to raise.
Thank you for the review!